### PR TITLE
Add metrics for Query Count statistics

### PR DESCRIPTION
**Pull request by nishantmonu51**
This PR adds a new metrics monitor “QueryCountStatsMonitor” which emits
these metrics (disabled by default) -
1) query/success/count - number of successful queries
2) query/failed/count - number of failed queries
3) query/interrupted/count - number of interrupted/timedout queries

### CODE REVIEW

File: server/src/main/java/io/druid/server/QueryResource.java [hunk 1]
```diff
@@ -94,6 +95,9 @@
   private final RequestLogger requestLogger;
   private final QueryManager queryManager;
   private final AuthConfig authConfig;
+  private volatile long successfulQueryCount = 0L;
```
Comment (leventov) [inline server/src/main/java/io/druid/server/QueryResource.java]: What's the point for `volatile` here?

File: server/src/main/java/io/druid/server/metrics/QueryCountStatsProvider.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.druid.server.metrics;
+
+public interface QueryCountStatsProvider
```
Comment (leventov) [inline server/src/main/java/io/druid/server/metrics/QueryCountStatsProvider.java]: What's the point for making interface until there is only one implementation?

File: server/src/main/java/io/druid/server/metrics/QueryCountStatsMonitor.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  )
+  {
+    this.statsProvider = statsProvider;
+  }
+
+  @Override
+  public boolean doMonitor(ServiceEmitter emitter)
+  {
+    final ServiceMetricEvent.Builder builder = new ServiceMetricEvent.Builder();
+    Map<String, Long> diff = keyedDiff.to(
```
Comment (leventov) [inline server/src/main/java/io/druid/server/metrics/QueryCountStatsMonitor.java]: `KeyedDiff` seems to me to be an overkill here (actually, to me, it seems to be an overkill everywhere, IMO this is just a pointless class that only complicates code). IMO it's simpler to just keep three fields with counters.

File: server/src/main/java/io/druid/server/QueryResource.java [hunk 4]
```diff
@@ -257,7 +261,7 @@ public void write(OutputStream outputStream) throws IOException, WebApplicationE
 
                     os.flush(); // Some types of OutputStream suppress flush errors in the .close() method.
                     os.close();
-
+                    successfulQueryCount++;
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/QueryResource.java]: Is `++` guaranteed atomic? My understanding is that this is not guaranteed to atomic across multiple threads calling it at the same time. Unless you can tell for sure that `++` is java guaranteed atomic on a volatile, then it would be safer to use an `AtomicLong`

File: server/src/main/java/io/druid/server/metrics/QueryCountStatsProvider.java (hunk 2 above)
Comment (drcrallen) [inline server/src/main/java/io/druid/server/metrics/QueryCountStatsProvider.java]: I'm going to guess guava

**Comment by fjy** [comment]
@nishantmonu51 there's a few conflicts

**Comment by b-slim** [comment]
@nishantmonu51 let's finish up this.

File: server/src/main/java/io/druid/server/QueryResource.java (hunk 4 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/server/QueryResource.java]: fixed.

File: server/src/main/java/io/druid/server/metrics/QueryCountStatsProvider.java (hunk 2 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/server/metrics/QueryCountStatsProvider.java]: IMHO, it makes code more clean in terms of monitor to pass a provider that exposes the required methods instead of passing complete object.

File: server/src/main/java/io/druid/server/metrics/QueryCountStatsMonitor.java (hunk 3 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/server/metrics/QueryCountStatsMonitor.java]: I have mixed opinion here, as the number of metrics increases with time the complexity with keeping individual fields also increases and chances of errors in making diff also increases.
Anyways, If you think we should simplify Keydiff usage everywhere, we can do that in a separate PR.

**Comment by nishantmonu51** [comment]
@leventov: not every druid user has a metric processing cluster to aggregate emitted metrics.

**Comment by nishantmonu51** [comment]
@b-slim @drcrallen: handled review comments.

File: server/src/main/java/io/druid/server/metrics/QueryCountStatsProvider.java (hunk 2 above)
Comment (leventov) [inline server/src/main/java/io/druid/server/metrics/QueryCountStatsProvider.java]: @nishantmonu51 if you are planning to make more implementations it is ok, but otherwise IMO it just makes the codebase more scattered, harder to find the meaning out of it and spawning questions (I asked this question, so probably other people who will read this code will also wonder are there any more implementations planned, try to find them, maybe in extensions, etc.)

Also given the limited usage of QueryCountStatsProvider: only within a single screen in a single class, there is no risk that "someone" will misuse `QueryResource`and call `doPost()` instead of `getSuccessfulQueryCount()`.

File: server/src/main/java/io/druid/server/metrics/QueryCountStatsMonitor.java (hunk 3 above)
Comment (leventov) [inline server/src/main/java/io/druid/server/metrics/QueryCountStatsMonitor.java]: ok

File: docs/content/operations/metrics.md [hunk 5]
```diff
@@ -34,6 +34,9 @@ Available Metrics
 |`query/node/bytes`|number of bytes returned from querying individual historical/realtime nodes.|id, status, server.| |
 |`query/node/ttfb`|Time to first byte. Milliseconds elapsed until broker starts receiving the response from individual historical/realtime nodes.|id, status, server.|< 1s|
 |`query/intervalChunk/time`|Only emitted if interval chunking is enabled. Milliseconds required to query an interval chunk.|id, status, chunkInterval (if interval chunking is enabled).|< 1s|
+|`query/success/count`|number of queries successfully processed|This metric is only available if the QueryCountStatsMonitor module is included.||
```
Comment (fjy) [inline docs/content/operations/metrics.md]: i dont see a querycountstats extension, only a module that is always included with query nodes

Comment (nishantmonu51) [inline docs/content/operations/metrics.md]: its not implemented as an extension.

File: server/src/main/java/io/druid/server/metrics/QueryCountStatsProvider.java (hunk 2 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/server/metrics/QueryCountStatsProvider.java]: while working on removing this, noticed that the router also needs to implement this. have modified router also to use it.

**Comment by nishantmonu51** [comment]
@leventov @b-slim: Please check again, all review comments are handled.

**Comment by b-slim** [comment]
👍

**Comment by fjy** [comment]
👍
